tests/test_parquet.py

```diff
@@ -771,12 +773,69 @@ def testBadDataFrameColumnParquet(self):
     def testWriteReadAstropyTableLossless(self):
         tab1 = _makeSimpleAstropyTable(include_multidim=True, include_masked=True)
 
-        self.butler.put(tab1, self.datasetType, dataId={})
+        put_ref = self.butler.put(tab1, self.datasetType, dataId={})
 
         tab2 = self.butler.get(self.datasetType, dataId={}, storageClass="ArrowAstropy")
 
+        # Check that minimal provenance was written by default.
+        expected = {
+            "lsst.butler.id": str(put_ref.id),
+            "lsst.butler.run": "test_run",
+            "lsst.butler.datasettype": "data",
+        }
+
+        self.assertEqual(tab2.meta, expected)
+
         _checkAstropyTableEquality(tab1, tab2)
 
+    @unittest.skipUnless(atable is not None, "Cannot test reading as astropy without astropy.")
+    def testWriteReadAstropyTableProvenance(self):
+        tab1 = _makeSimpleAstropyTable()
+
+        # Create a ref for provenance.
+        astropy_type = DatasetType(
+            "astropy_parquet",
+            dimensions=(),
+            storageClass="ArrowAstropy",
+            universe=self.butler.dimensions,
+        )
+        self.butler.registry.registerDatasetType(astropy_type)
+        input_ref = DatasetRef(astropy_type, {}, run="other_run")
+        quantum_id = uuid.uuid4()
+        provenance = DatasetProvenance(quantum_id=quantum_id)
+        provenance.add_input(input_ref)
+
+        put_ref = self.butler.put(tab1, self.datasetType, dataId={}, provenance=provenance)
+
+        tab2 = self.butler.get(self.datasetType, dataId={}, storageClass="ArrowAstropy")
+
+        expected = {
+            "lsst.butler.id": str(put_ref.id),
+            "lsst.butler.run": "test_run",
+            "lsst.butler.datasettype": "data",
+            "lsst.butler.quantum": str(quantum_id),
+            "lsst.butler.input.0.id": str(input_ref.id),
+            "lsst.butler.input.0.run": "other_run",
+            "lsst.butler.input.0.datasettype": "astropy_parquet",
+        }
+        self.assertEqual(tab2.meta, expected)
+
+        # Put the dataset again, with different provenance and ensure
+        # that the previous provenance was stripped.
+        self.butler.collections.register("new_run")
+        put_ref3 = self.butler.put(tab2, self.datasetType, dataId={}, run="new_run")
+
+        # tab2 will have been updated in place.
+        expected = {
+            "lsst.butler.id": str(put_ref3.id),
+            "lsst.butler.run": "new_run",
+            "lsst.butler.datasettype": "data",
+        }
+        self.assertEqual(tab2.meta, expected)
+        null_prov, prov_ref = DatasetProvenance.from_flat_dict(tab2.meta, self.butler)
```

Comment:
So if anybody were to set a metadata key that was `lsst.butler.something` would this break?  Just curious; it seems fair to have these as reserved keys but I don't know how we could enforce that.

Reply:
It seems fair to me that only butler can set `lsst.butler.` keys (that's the point of the namespacing). If someone added a `lsst.butler.input 5 id` or something then this would break because the separators would be confused. I think so long as `something` was not `run` or `id` or `datasettype` or `input` then this code wouldn't notice.